src/libaktualizr/utilities/dequeue_buffer.h

```diff
@@ -0,0 +1,55 @@
+#ifndef UPTANE_DEQUEUE_BUFFER_H_
+#define UPTANE_DEQUEUE_BUFFER_H_
+
+#include <cstddef>
+/**
+ * A dequeue based on a contiguous buffer in memory. Used for buffering
+ * data between recv() and ber_decode()
+ */
+class DequeueBuffer {
+ public:
+  /**
+   * A pointer to the first element that has not been Consumed().
+   */
+  char *Head();
+
+  /**
+   * The number of elements that are valid (have been written) after Head()
+   */
+  size_t Size();
+
+  /**
+   * Called after bytes have been read from Head(). Remove them from the head
+   * of the queue.
+   */
+  void Consume(size_t bytes);
+
+  /**
+   * A pointer to the next place to write data to
+   */
+  char *Tail();
+
+  /**
+   * The number of bytes beyond Tail() that are allocated and may be written to.
+   */
+  size_t TailSpace();
+
+  /**
+   * Call to indicate that bytes have been written in the range
+   * Tail() ... Tail() + TailSpace().
+   */
+  void HaveEnqueued(size_t bytes);
+
+ private:
+  static constexpr size_t kBufferSize = 4096;
+  static constexpr int kSentinel = 1337161494;
+
+  /**
+   * buffer_[0..written_bytes_] contains to contents of this dequeue
+   */
+  size_t written_bytes_{0};
+  char buffer_[kBufferSize]{};  // Zero initialise as a security pesimisation
+  int sentinel_{kSentinel};     // Sentinel to check for writers overflowing buffer_
```

Comment:
With the current level of optimization going on, not sure if we can rely on this... I'd say it's not worth the complication.
Plus all calls to assert are optimized out if `NDEBUG` is not set: http://en.cppreference.com/w/cpp/error/assert

Instead, it could maybe use a `std::array` with bound-checked accessors: http://en.cppreference.com/w/cpp/container/array/at

Reply:
Still, switching to std::array makes sense. It shouldn't be hard

Reply:
Then it may be worth checking `span`, maybe it'd apply here: http://codexpert.ro/blog/2016/03/07/guidelines-support-library-review-spant/ (standardized in C++20, but also available in some externally supported libraries).

Maybe that's too future-y :)

Reply:
Span looks interesting, however the goal of DequeueBuffer is to provide an API that fits between `recv()` and `ber_decode()`, so the external interface needs to be based on raw pointers and lengths.  I can't see how using it internally would help the implementation (which is basically just a memmove (now std::copy_backwards) when data is consumed off the head.